teps/0146-parameters-in-script.md

```diff
@@ -0,0 +1,605 @@
+    (e.g. the `git-cli` Task [accepts a script to run as a param][git-cli]) can be updated
+    to execute the contents of the `Parameter` from a file or call `eval` on it explicitly.
+
+3. Tekton Pipelines and Catalog documentation should be updated to the new approach.
+
+4. **`Tasks` should not support implicit runtime code injection**.
+
+   Disallow interpolating `Param` values directly into executable
+   contexts. If the user wants to execute the contents of a param as a
+   script then they should be able to, but it shouldn't be possible
+   without "opting in" to that behaviour by calling `eval` or writing
+   the value of a param to a script and running it.
+
+#### How will we know that this has succeeded?
+We will now this has succeeded:
+  1. When there is a method of securely and reliably uses `Parameters` in `script`s 
+  2. When the current method with less security and reliability is deprecated/no-longer-allowed and using `Parameters` in `script`s is secure by default
+
+### Non-Goals
+Directly disallowing the use of `Param` values in `script`s without providing a feasible alternative.
+
+### Use Cases
+As described in the [original proposal][pipeline-issue-781], the use case for `Script`
+in `Tasks` is:
+
+> Users need an easy cruft-free way to express a multi-statement `Script` in the
+> body of a `Task` `Step`, without having to understand advanced topics about
+> `Containers`, like what an `entrypoint` or a `command` is.
+>
+> Users should only need to have passing familiarity with a shell environment
+> to be successful using Tekton.
+
+Problems with variable subsitution was noted in [original proposal][pipeline-issue-781]:
+
+> How this syntax works with input substitutions, e.g., ``${inputs.params.flags}`` might
+> also lead to confusion among users, which we should design for.
+
+Users need a way to safely and reliably substitute `Parameters` in `Scripts` in `Steps`
+of their `Tasks`. This would make it easier to use Tekton without learning advanced topics,
+while still using it reliably and safely.
+
+### Requirements
+
+<!--
+Describe constraints on the solution that must be met, such as:
+- which performance characteristics that must be met?
+- which specific edge cases that must be handled?
+- which user scenarios that will be affected and must be accommodated?
+-->
+
+1. **[Secure by default in v1][secure-by-default]**: Out of the box
+   *Tekton Pipelines* v1 should disallow common anti-patterns that
+   can render a default install insecure.
+1. **Easy to use**: Whatever replaces the direct use of `$(params.X)` in
+   `script` should be unsurprising and memorable to ease friction in user
+   migration.
+1. **Automatable**: A script or program should be able to convert `Tasks`
+   to the new approach. This might be impossible for all cases (e.g. if an
+   unrecognized shebang has been used in the script) but a best-effort
+   implementation should be provided.
```

Comment:
It would be nice to have such tooling indeed, maybe something that can detect issues and optionally attempt to fix them.

Reply:
+1, as part of the implementation we would add some automation that did either the following:
* Either as a warning in controller logs OR when the insecure defaults are eventually removed (via opt-in feature flag or after decided upon depreciation timing in v1.X, v2, etc.) - tekton will suggests running a new to-be-implemented CLI command such as `tkn fix` that could update input YAML files, converting them to using the new secure “.env” syntax for `script` and relaying any issues/manual changes necessary as needed
* Controller implementation that is able to convert all/subset of cases the automatically “under the hood”.  Ideally this would work for all cases but perhaps for some 80% (eg: bash only, etc.) cases it works and for others it returns a validation error